service/src/BatchServer.cpp

```diff
@@ -0,0 +1,250 @@
+ */
+
+#include "config.h"
+#include "BatchServer.hpp"
+
+#include <cstdlib>
+#include <cerrno>
+#include <unistd.h>
+
+#include "geopm/Exception.hpp"
+#include "geopm/PlatformIO.hpp"
+#include "geopm/SharedMemory.hpp"
+#include "geopm/Helper.hpp"
+#include "geopm/PlatformIO.hpp"
+#include "BatchStatus.hpp"
+#include "geopm_debug.hpp"
+
+
+namespace geopm
+{
+    std::unique_ptr<BatchServer>
+    BatchServer::make_unique(int client_pid,
+                             const std::vector<geopm_request_s> &signal_config,
+                             const std::vector<geopm_request_s> &control_config)
+    {
+        return geopm::make_unique<BatchServerImp>(client_pid, signal_config,
+                                                  control_config);
+    }
+    BatchServerImp::BatchServerImp(int client_pid,
+                                   const std::vector<geopm_request_s> &signal_config,
+                                   const std::vector<geopm_request_s> &control_config)
+        : BatchServerImp(client_pid, signal_config, control_config,
+                         platform_io(), nullptr, nullptr, nullptr)
+    {
+        // Fork the server when calling real constructor.
+        auto setup = [this]() {this->create_shmem();};
+        auto run = [this]() {this->run_batch();};
+        m_server_pid = fork_with_setup(setup, run);
+    }
+
+    BatchServerImp::BatchServerImp(int client_pid,
+                                   const std::vector<geopm_request_s> &signal_config,
+                                   const std::vector<geopm_request_s> &control_config,
+                                   PlatformIO &pio,
+                                   std::shared_ptr<BatchStatus> batch_status,
+                                   std::shared_ptr<SharedMemory> signal_shmem,
+                                   std::shared_ptr<SharedMemory> control_shmem)
+        : m_client_pid(client_pid)
+        , m_server_key(std::to_string(m_client_pid))
+        , m_signal_config(signal_config)
+        , m_control_config(control_config)
+        , m_pio(pio)
+        , m_signal_shmem(signal_shmem)
+        , m_control_shmem(control_shmem)
+        , m_batch_status(batch_status != nullptr ?
+                         batch_status : BatchStatus::make_unique_server(m_client_pid,
+                                                                        m_server_key))
+        , m_server_pid(0)
+        , m_is_active(true)
+    {
+
+    }
+
+    BatchServerImp::~BatchServerImp()
+    {
+        stop_batch();
+    }
+
+    int BatchServerImp::server_pid(void) const
+    {
+        return m_server_pid;
+    }
+
+    std::string BatchServerImp::server_key(void) const
+    {
+        return m_server_key;
+    }
+
+    void BatchServerImp::stop_batch(void)
+    {
+        if (m_is_active) {
+            m_batch_status->send_message(M_MESSAGE_QUIT);
+            m_batch_status->receive_message(M_MESSAGE_CONTINUE);
+            m_is_active = false;
```

Comment:
This message is going the wrong way.  The caller of this method is the parent process that forked the batch server (i.e. the primary geopmd process).  It doesn't make sense for geopmd to send a message to the client process to "quit".  This should be probably be resolved with a signal sent by geopmd and a signal handler on the batch server process.